erts/emulator/beam/erl_gc.c

```diff
@@ -3839,6 +3839,173 @@ erts_max_heap_size(Eterm arg, Uint *max_heap_size, Uint *max_heap_flags)
+            break;
+        case TAG_PRIMARY_HEADER:
+            if (header_is_transparent(val)) {
+                continue;
+            }
+            switch (thing_subtag(val)) {
+                case SUB_BITS_SUBTAG:
+                {
+                    const ErlSubBits *sb = (ErlSubBits*)(tp-1);
+                    const BinRef *underlying = (BinRef*)boxed_val(sb->orig);
+                    if (thing_subtag(underlying->thing_word) != HEAP_BITS_SUBTAG) {
+                        for (Uint i = 0; i < count; i++) {
+                            DebugBinary* b = &bins[i];
+                            if (b->bin_ref == underlying) {
+                                WSTACK_PUSH2(b->ws.ws, sb->start, sb->end);
+                                break;
+                            }
+                        }
+                    }
+                }
+                ERTS_FALLTHROUGH();
+            default:
+                {
+                    tp += header_arity(val);
+                }
+                break;
+            }
+            break;
+        }
+    }
+}
+
+Eterm
+erts_gather_binaries(ErtsHeapFactory *hfact, Process *rp) {
+    #define PSTACK_TYPE DebugBinary
+    PSTACK_DECLARE(binaries, 16);
+
+    union erl_off_heap_ptr u;
+    Eterm res = NIL;
+    Eterm tuple;
+    union erts_tmp_aligned_offheap tmp;
+    Uint binaries_count;
+    DebugBinary* binariesp;
+
+    ErlHeapFragment* bp;
+    ErtsMessage* mp;
+    Eterm *htop, *heap;
+    Uint sz = 0;
+    Eterm *hp;
+
+    for (u.hdr = MSO(rp).first; u.hdr; u.hdr = u.hdr->next) {
+        erts_align_offheap(&u, &tmp);
+        if (u.hdr->thing_word == HEADER_BIN_REF) {
+            DebugBinary* bin = PSTACK_PUSH(binaries);
+            bin->bin_ref = u.br;
+            WSTACK_INIT(&bin->ws, ERTS_ALC_T_ESTACK);
+        }
+    }
+
+    for (u.hdr = rp->wrt_bins; u.hdr; u.hdr = u.hdr->next) {
+        erts_align_offheap(&u, &tmp);
+        if (u.hdr->thing_word == HEADER_BIN_REF) {
+            DebugBinary* bin = PSTACK_PUSH(binaries);
+            bin->bin_ref = u.br;
+            WSTACK_INIT(&bin->ws, ERTS_ALC_T_ESTACK);
+        }
+    }
+    binariesp = (DebugBinary*)binaries.pstart;
+    binaries_count = PSTACK_COUNT(binaries);
+
+    if (rp->abandoned_heap) {
+        heap = get_orig_heap(rp, &htop, NULL);
+        gather_binaries(binariesp, binaries_count, heap, htop);
+    }
+
+    if (OLD_HEAP(rp))
+        gather_binaries(binariesp, binaries_count, OLD_HEAP(rp), OLD_HTOP(rp) /*OLD_HEND(p)*/);
+
+    gather_binaries(binariesp, binaries_count, HEAP_START(rp), HEAP_TOP(rp));
+
+    mp = rp->msg_frag;
+    bp = rp->mbuf;
+
+    if (bp)
+        goto search_heap_frags;
+
+    while (mp) {
+
+        bp = erts_message_to_heap_frag(mp);
+        mp = mp->next;
+
+    search_heap_frags:
+
+        while (bp) {
+            gather_binaries(binariesp, binaries_count,
+                bp->mem, bp->mem + bp->used_size);
+            bp = bp->next;
+        }
+    }
```

Comment:
`mbuf` and `msg_frag` need to be searched separately, as you can have heap fragments without messages. Both have `next` members that you should follow, and the latter should keep using `erts_message_to_heap_frag` like you do now.